### PR TITLE
[onDeviceUI] Add ability to render addons in React Native

### PR DESCRIPTION
**Pull request by Gongreg**
# Issue 
Currently the only way to use addons in React Native is to display them in browser (that means you need to have a server running). This PR allows users to create addons which work in RN app itself.

You can test out the new functionality here! https://expo.io/@gongreg/crna-kitchen-sink.

40x

This is the main PRs for this functionality. Relevant PRs: 

#4326
#4327
#4328
#4340

## Changes

* #### Updated the UI.
  * Moved top bar to the bottom, added border to make it more distinct.
  * Bar now has three buttons: Navigator, preview, addons. Clicking on either navigator or addons makes the preview minimized.
  * You can swipe on the bar to select different panels.
  * You can press on minimized preview to maximize it.
  * Added a button on the right to hide the bar completely.

* #### `onDeviceUI` is now true by default.
You can disable it by passing onDeviceUI as false.

* #### Two new options for initial UI display
  *  isUIOpen - display bottom bar initially (default - true).
  *  tabOpen - which panel to display initially (navigator(-1), preview(0) or addons(1) (default is 0).

* #### Safe area view hack (for iphone x) removed.
Now it uses SafeAreaView component. It requires RN 0.50.

* #### Reduced amount of rerenders happening when using onDeviceUI.
Before using onDeviceUI made two rerenders on story change events.

* #### Channel is created immediately.
Before it was impossible to use addons channel if you didn't render storybookUI component. Now the channel is created as soon as you call getStorybookUI.

## Testing
I've updated CRNA example to support on device addons.
I've also published expo example: https://expo.io/@gongreg/crna-kitchen-sink

## Part of other PR's

* #### Addon registration
cli now creates rn-addons.js file and imports it in entry storybook file.

Current storybook architecture works in two parts. Manager and preview. Adding addons to onDeviceUI removes this separation. 

This separation was necessary to support emitting events inside decorators.
Without it now we get warning that renders are not "pure" (they are calling setState while setState is already happening).

By using async channel we avoid this issue since events are emitted only after render has finished.

* #### Handles websocket error without crashing.
If storybook tries to connect to non existing storybook server it will not throw red screen anymore.

* #### Created new on device addons
  * Backgrounds
  * Knobs
  * Notes

* #### Unified template for Storybook installation in React Native.
Both CRNA and RN Vanilla will use same template when installing storybook.

### CODE REVIEW

**Comment by orta** [comment]
This is a very clever idea 👍

File: app/react-native/src/index.js [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import { AddonStore } from '@storybook/addons';
```
Comment (igor-dv) [inline app/react-native/src/index.js]: > Web part of storybook uses addons.js file. Maybe we should have rn-addons.js file for RN part? Currently I just register the addons next to the stories.

I think having `addon.js` for RN will consolidate the API.

File: examples/crna-kitchen-sink/storybook/index.js [hunk 2]
```diff
@@ -3,12 +3,28 @@ import React, { Component } from 'react';
 import { getStorybookUI, configure } from '@storybook/react-native';
 import { setOptions } from '@storybook/addon-options';
 
+// TODO temp disable till I find a way to fix usage addon
+console.disableYellowBox = true;
```
Comment (shilman) [inline examples/crna-kitchen-sink/storybook/index.js]: WHOA 🙀

**Comment by ndelangen** [comment]
@Gongreg show them the latest iteration on the idea ;)

**Comment by Gongreg** [comment]
Hey. @ndelangen had a chance to try out my PR. It felt a little bit clunky for him, so we've decided to update the UI. This is just a POC, so don't mind all those bright colors.

What we've decided to do is to display navigator/addons as panels from the side while the preview is minimized. So you can still see what is happening in the preview, but you have place to work with.

giphy

File: app/react-native/docs/addons.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+## Compactibility
+Addon compatibilty can be found [here](https://github.com/storybooks/storybook/blob/master/ADDONS_SUPPORT.md)
+
+## Performance of app addons
+There is one big difference in performance of web addons and app addons. Because app addons are inside the app, they are
+also rerendered on every change. If you feel any performance implications, this may be a cause.
+ 
+## Writing the app addons
+App addons have same api as browser addons except currently api prop is not supported in React Native (you can use channel for most problems).
```
Comment (Gongreg) [inline app/react-native/docs/addons.md]: Unclear sentence

File: app/react-native/docs/manual-setup.md [hunk 4]
```diff
@@ -10,22 +10,42 @@ Create a new directory called `storybook` in your project root and create an ent
...
+  // eslint-disable-next-line global-require
   require('./stories');
 }, module);
 
-const StorybookUI = getStorybookUI({
-  port: 7007,
-  host: 'localhost',
-});
-AppRegistry.registerComponent('MyApplicationName', () => StorybookUI);
+const StorybookUIRoot = getStorybookUI({ port: 7007, onDeviceUI: true });
```
Comment (Gongreg) [inline app/react-native/docs/manual-setup.md]: just getStorybookUI()

File: app/react-native/docs/manual-setup.md [hunk 5]
```diff
@@ -10,22 +10,42 @@ Create a new directory called `storybook` in your project root and create an ent
...
+AppRegistry.registerComponent('MyApplicationName', () => StorybookUIHMRRoot);
+export default StorybookUIHMRRoot;
 ```
 
-Create a file named `addons.js` file in `storybook` directory to use addons. Here is a list of default addons:
+Create a file called rn-addons.js
+
+In this file you can import all the addons that work inside rn simulator.
+
+If you want to have addons inside browser, create a file named `addons.js` file in `storybook`. Here is a list of default addons:
```
Comment (Gongreg) [inline app/react-native/docs/manual-setup.md]: Remove default addons text

File: app/react-native/readme.md [hunk 6]
```diff
@@ -126,8 +126,8 @@ The following parameters can be passed to the start command:
     DEVELOPMENT/PRODUCTION environment for webpack
 -r, --reset-cache
     reset react native packager
---skip-packager
-    run only storybook server
+--start-packager
```
Comment (Gongreg) [inline app/react-native/readme.md]: Deprecated

File: examples/crna-kitchen-sink/storybook/stories/Knobs/index.js [hunk 7]
```diff
@@ -23,12 +23,12 @@ export default () => {
   };
   const fruit = select('Fruit', fruits, 'apple');
 
-  const otherFruits = {
-    Kiwi: 'kiwi',
-    Guava: 'guava',
-    Watermelon: 'watermelon',
-  };
-  const otherFruit = radios('Other Fruit', otherFruits, 'watermelon');
+  //const otherFruits = {
```
Comment (Gongreg) [inline examples/crna-kitchen-sink/storybook/stories/Knobs/index.js]: Add radios support.

File: lib/cli/generators/REACT_NATIVE_SCRIPTS/template/storybook/rn-addons.js [hunk 8]
```diff
@@ -0,0 +1,2 @@
+import '@storybook/addon-actions/register';
```
Comment (Gongreg) [inline lib/cli/generators/REACT_NATIVE_SCRIPTS/template/storybook/rn-addons.js]: Remove these two addons, they don't work in rn yet

**Comment by Gongreg** [comment]
If you want to test thing out I've create expo example: https://expo.io/@gongreg/crna-kitchen-sink

**Comment by Gongreg** [comment]
I've fixed most of the issues with Android, so if you want to, you can play with it.
Feedback is very welcome.

File: app/react-native/docs/server.md [hunk 9]
```diff
@@ -0,0 +1,23 @@
+# Storybook server
+The default usage of React native storybook till version 5 involved starting storybook server.
```
Comment (igor-dv) [inline app/react-native/docs/server.md]: version 5 ? 😄

Comment (Gongreg) [reply app/react-native/docs/server.md]: Thanks. Good catch

File: app/react-native/docs/addons.md [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+There is one big difference in React Native is that it has two types of addons: Addons that work in the browser
+and addons that work on the app itself (on device adons).
+
+## Browser addons
+Browser addons are default addons to storybook. You create a file called addons.js inside storybook and it is
+automatically added inside your browser.
+
+## On device addons
+On device addons are addons that are displayed in your app in addons panel. You register them in a file called 
+`rn-addons.js` and you also have to import `rn-addons.js` file somewhere in your entry file (it is done by default
```
Comment (Gongreg) [inline app/react-native/docs/addons.md]: Needs better usage explanation

File: app/react-native/docs/addons.md [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+## Compactibility
+Addon compatibilty can be found [here](https://github.com/storybooks/storybook/blob/master/ADDONS_SUPPORT.md)
+
+## Performance of on device addons
+There is one big difference in performance of web addons and on device addons. Because on device addons are inside the 
+app, they are also rerendered on every change.
+ 
+## Writing the app addons
+On device addons use same addon store and api as web addons. The only difference in api is that you don't have api prop
```
Comment (Gongreg) [inline app/react-native/docs/addons.md]: api prop in quotes

File: app/react-native/docs/manual-setup.md [hunk 12]
```diff
@@ -6,32 +6,45 @@ First, install the `@storybook/react-native` module
 npm install @storybook/react-native
```
Comment (Gongreg) [inline app/react-native/docs/manual-setup.md]: Update this doc

File: app/react-native/docs/server.md [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+* ### Story selection
+Having server running allows you to control your storybook view from inside web page or your ide.
+
+There is a plugin for [JetBrains IDEs](https://plugins.jetbrains.com/plugin/9910-storybook) and there is one 
+for [VS Code](https://github.com/orta/vscode-react-native-storybooks)
+
+
+* ### Web addons
+There are addons that work on React Native but don't have on device implementation yet.
```
Comment (Gongreg) [inline app/react-native/docs/server.md]: And some addons simply make sense only on the web.

File: app/react-native/docs/server.md [hunk 14]
```diff
@@ -0,0 +1,23 @@
+# Storybook server
+The default usage of React native storybook till version 5 involved starting storybook server.
+
+The reason it has changed is that in most cases it is not really necessary.
```
Comment (Gongreg) [inline app/react-native/docs/server.md]: Explain more here.

File: app/react-native/src/preview/index.js [hunk 15]
```diff
@@ -40,50 +40,85 @@ export default class Preview {
...
+    // If the app is started with server running, set the story as the one selected in the browser
+    if (webUrl) {
       this._sendGetCurrentStory();
+    } else {
+      setInitialStory = true;
+    }
+
+    const preview = this;
+
+    // react-native hot module loader must take in a Class - https://github.com/facebook/react-native/issues/10991
```
[the remark this answers is not in the document]
Comment (Gongreg) [reply app/react-native/src/preview/index.js]: According to one of the comments it says that from RN 52 it should be fixed. But I am not really sure, we can remove it after.

**Review by ndelangen — APPROVED** [review]

**Review by igor-dv — APPROVED** [review]

**Comment by Gongreg** [comment]
This can move forward! @ndelangen, @igor-dv.

**Comment by slorber** [comment]
hey,

This works great, but just wondering if it wouldn't be possible to attach plugins to a chrome extension? The app can already connect to chrome debugger so maybe we could also support showing plugin interfaces like knobs here without needing a server?

**Comment by Gongreg** [comment]
Hey @slorber, I don't have enough knowledge with debugger to comment on that one. But if you have any interest in trying to do that you are definitely welcome.

**Pull request merged**